### PR TITLE
Implement mTLS with test

### PR DESCRIPTION
**Pull request by MykolaSolyanko**

### CODE REVIEW

File: src/communication/tlsvchannel.hpp [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+     * Writes data from array to channel.
+     *
+     * @param data array where data is taken from.
+     * @param size specifies how many bytes to write.
+     * @return aos::Error.
+     */
+    aos::Error Write(const aos::Array<uint8_t>& data) override;
+
+private:
+    static constexpr const char* cPers = "tls_vchannel_client";
```
Comment (al1img) [inline src/communication/tlsvchannel.hpp:80]: `static constexpr auto` ?

File: src/communication/commchannel.hpp [hunk 2]
```diff
@@ -26,6 +26,7 @@ class CommChannelItf {
     /**
      * Connects to communication channel.
      *
+     * @param secure specifies if secure connection is required.
```
Comment (al1img) [inline src/communication/commchannel.hpp:29]: Remove this comment.

File: src/communication/tlschannel.hpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+     *
+     * @param certHandler certificate handler.
+     * @param certLoader certificate loader.
+     * @param vChannel virtual channel.
+     */
+    aos::Error Init(aos::iam::certhandler::CertHandlerItf& certHandler, aos::cryptoutils::CertLoaderItf& certLoader,
+        CommChannelItf& vChannel);
+
+    /**
+     * Set the secure channel.
```
Comment (al1img) [inline src/communication/tlschannel.hpp:40]: Set TLS config

File: src/communication/tlschannel.hpp [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+     *
+     * @param certType Certificate type.
+     * @return aos::Error
+     */
+    aos::Error SetTLSConfig(const aos::String& certType) override;
+
+    /**
+     * Connects to communication channel.
+     *
+     * @param secure specifies if secure connection is required.
```
Comment (al1img) [inline src/communication/tlschannel.hpp:50]: remove param

File: src/communication/tlschannel.hpp [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+     * Returns if channel is connected.
+     *
+     * @return bool.
+     */
+    bool IsConnected() const override;
+
+    /**
+     * Reads data from channel to array.
+     *
+     * @param data array where data is placed to.
```
Comment (al1img) [inline src/communication/tlschannel.hpp:70]: Copy comments from  CommChannelItf

File: src/communication/tlschannel.hpp [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+     * @param size specifies how many bytes to read.
+     * @return aos::Error.
+     */
+    int Read(void* data, size_t size) override;
+
+    /**
+     * Writes data from array to channel.
+     *
+     * @param data array where data is taken from.
```
Comment (al1img) [inline src/communication/tlschannel.hpp:79]: ditto

File: src/communication/vchannel.hpp [hunk 7]
```diff
@@ -41,6 +41,7 @@ class VChannel : public CommChannelItf {
     /**
      * Connects to communication channel.
      *
+     * @param secure specifies if secure connection is required.
```
Comment (al1img) [inline src/communication/vchannel.hpp:44]: Remove

File: tests/communication/src/mocks/commchannelmock.hpp [hunk 8]
```diff
@@ -12,6 +12,8 @@
 #include <mutex>
 #include <vector>
 
+#include <aos/common/tools/string.hpp>
```
Comment (al1img) [inline tests/communication/src/mocks/commchannelmock.hpp:15]: Is it required?

File: tests/communication/CMakeLists.txt [hunk 9]
```diff
@@ -8,10 +8,41 @@ find_package(
...
+
 set(CMAKE_MODULE_PATH ${APPLICATION_SOURCE_DIR}/../../CMakeModules)
 
 project(communication_test)
 
+# ######################################################################################################################
+# Certificates
+# ######################################################################################################################
+
+set(CERTDIR ${CMAKE_CURRENT_BINARY_DIR}/../../../certificates)
```
Comment (al1img) [inline tests/communication/CMakeLists.txt:21]: It should be inside binary dir

File: tests/communication/CMakeLists.txt [hunk 10]
```diff
@@ -51,13 +82,43 @@ target_include_directories(app PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
 # Target
 # ######################################################################################################################
 
+# Enable AESNI for posix 32bit
+target_compile_options(mbedTLS INTERFACE -mpclmul -msse2 -maes)
+
+set(aoscore_source_dir "${CMAKE_CURRENT_SOURCE_DIR}/../../../aos_core_lib_cpp")
```
Comment (al1img) [inline tests/communication/CMakeLists.txt:88]: Move it to top and use in other places where aos_core_lib_cpp is referenced

File: tests/communication/CMakeLists.txt [hunk 11]
```diff
@@ -51,13 +82,43 @@ target_include_directories(app PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
...
+
+target_compile_definitions(app PRIVATE CERT_DIR="${CERTDIR}")
+
 target_sources(
     app
     PRIVATE ../../src/communication/communication.cpp
             ../../src/communication/cmclient.cpp
+            ../../src/communication/tlschannel.cpp
+            ../../src/rootca/rootca.S
+            ${aoscore_source_dir}/src/common/crypto/mbedtls/driverwrapper.cpp
```
Comment (al1img) [inline tests/communication/CMakeLists.txt:98]: Put it to the bottom of list. Other groups sort by alphabet.

File: tests/communication/CMakeLists.txt [hunk 12]
```diff
@@ -51,13 +82,43 @@ target_include_directories(app PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
...
+    ${ZEPHYR_MBEDTLS_MODULE_DIR}/library/psa_crypto_driver_wrappers_no_static.c
+    ${ZEPHYR_MBEDTLS_MODULE_DIR}/library/psa_crypto.c TARGET_DIRECTORY mbedTLS PROPERTIES HEADER_FILE_ONLY ON
 )
+
+target_sources(
+    app PRIVATE ${aoscore_source_dir}/src/common/crypto/mbedtls/drivers/psa_crypto_driver_wrappers_no_static.c
+                ${aoscore_source_dir}/src/common/crypto/mbedtls/drivers/psa_crypto.c
+)
+
+message(STATUS "AOS CORE SOURCE DIR: ${aoscore_source_dir}")
```
Comment (al1img) [inline tests/communication/CMakeLists.txt:117]: remove

File: tests/communication/CMakeLists.txt [hunk 13]
```diff
@@ -51,13 +82,43 @@ target_include_directories(app PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
...
+
+set(CUSTOM_MBEDTLS_CONFIG_FILE ${aoscore_source_dir}/src/common/crypto/mbedtls/drivers/psa_crypto_driver_wrappers.h)
```
Comment (al1img) [inline tests/communication/CMakeLists.txt:119]: Why?

File: tests/communication/CMakeLists.txt [hunk 14]
```diff
@@ -51,13 +82,43 @@ target_include_directories(app PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
...
+
+configure_file(${CUSTOM_MBEDTLS_CONFIG_FILE} ${ZEPHYR_MBEDTLS_MODULE_DIR}/library/psa_crypto_driver_wrappers.h COPYONLY)
```
Comment (al1img) [inline tests/communication/CMakeLists.txt:121]: Could we have the same mbedtls configuration as in the main cmake?

File: tests/communication/src/tlschannel.cpp [hunk 15]
```diff
@@ -0,0 +1,455 @@
...
+ * Copyright (C) 2024 Renesas Electronics Corporation.
+ * Copyright (C) 2024 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/tc_util.h>
+#include <zephyr/ztest.h>
+
+#include <future>
```
Comment (al1img) [inline tests/communication/src/tlschannel.cpp:11]: system headers on the top

File: tests/communication/src/tlschannel.cpp [hunk 16]
```diff
@@ -0,0 +1,455 @@
...
+    {
+        return aos::ErrorEnum::eNotSupported;
+    }
+
+public:
+    aos::crypto::RSAPublicKey mPublicKey;
+    mbedtls_pk_context        mPkContext;
+};
+
+class MockCertLoader : public aos::cryptoutils::CertLoaderItf {
```
Comment (al1img) [inline tests/communication/src/tlschannel.cpp:88]: Move all mock classes into mock folder. And name them accordingly.

File: tests/communication/src/tlschannel.cpp [hunk 17]
```diff
@@ -0,0 +1,455 @@
...
+
+#include <aos/common/crypto.hpp>
+#include <aos/common/cryptoutils.hpp>
+#include <aos/common/tools/error.hpp>
+#include <aos/common/tools/fs.hpp>
+#include <aos/iam/certhandler.hpp>
+
+#include "communication/commchannel.hpp"
+#include "communication/tlschannel.hpp"
+
```
Comment (al1img) [inline tests/communication/src/tlschannel.cpp:29]: Add headers:
Types, Vars, Tests etc.

File: src/communication/commchannel.hpp (hunk 2 above)
Comment (MykolaSolyanko) [reply src/communication/commchannel.hpp:29]: Done

File: src/communication/tlschannel.hpp (hunk 3 above)
Comment (MykolaSolyanko) [reply src/communication/tlschannel.hpp:40]: Done

File: src/communication/tlschannel.hpp (hunk 4 above)
Comment (MykolaSolyanko) [reply src/communication/tlschannel.hpp:50]: Done

File: src/communication/tlschannel.hpp (hunk 5 above)
Comment (MykolaSolyanko) [reply src/communication/tlschannel.hpp:70]: Done

File: src/communication/tlschannel.hpp (hunk 6 above)
Comment (MykolaSolyanko) [reply src/communication/tlschannel.hpp:79]: Done

File: tests/communication/src/mocks/commchannelmock.hpp (hunk 8 above)
Comment (MykolaSolyanko) [reply tests/communication/src/mocks/commchannelmock.hpp:15]: no, removed

File: tests/communication/CMakeLists.txt (hunk 10 above)
Comment (MykolaSolyanko) [reply tests/communication/CMakeLists.txt:88]: Done

File: tests/communication/CMakeLists.txt (hunk 11 above)
Comment (MykolaSolyanko) [reply tests/communication/CMakeLists.txt:98]: Done

File: tests/communication/CMakeLists.txt (hunk 12 above)
Comment (MykolaSolyanko) [reply tests/communication/CMakeLists.txt:117]: Done

File: tests/communication/CMakeLists.txt (hunk 14 above)
Comment (MykolaSolyanko) [reply tests/communication/CMakeLists.txt:121]: removed

File: tests/communication/CMakeLists.txt (hunk 9 above)
Comment (MykolaSolyanko) [reply tests/communication/CMakeLists.txt:21]: Done

File: tests/communication/src/tlschannel.cpp (hunk 15 above)
Comment (MykolaSolyanko) [reply tests/communication/src/tlschannel.cpp:11]: Done

File: tests/communication/src/tlschannel.cpp (hunk 16 above)
Comment (MykolaSolyanko) [reply tests/communication/src/tlschannel.cpp:88]: Done

File: src/communication/tlschannel.hpp [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+     * @param certLoader certificate loader.
+     * @param vChannel virtual channel.
+     */
+    aos::Error Init(aos::iam::certhandler::CertHandlerItf& certHandler, aos::cryptoutils::CertLoaderItf& certLoader,
+        CommChannelItf& vChannel);
+
+    /**
+     * Set TLS config.
+     *
+     * @param certType Certificate type.
```
Comment (al1img) [inline src/communication/tlschannel.hpp:42]: small letter `certificate type.`

File: tests/communication/src/mocks/commchannelmock.hpp (hunk 8 above)
Comment (al1img) [reply tests/communication/src/mocks/commchannelmock.hpp]: Still there

File: tests/communication/CMakeLists.txt [hunk 19]
```diff
@@ -51,13 +83,35 @@ target_include_directories(app PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
...
+target_compile_options(mbedTLS INTERFACE -mpclmul -msse2 -maes)
+
+target_compile_definitions(app PRIVATE CERT_DIR="${CERTDIR}")
+target_compile_options(app PRIVATE -DCONFIG_AOS_ROOT_CA_PATH="${CERTDIR}/ca.pem")
 target_sources(
     app
     PRIVATE ../../src/communication/communication.cpp
             ../../src/communication/cmclient.cpp
+            ../../src/communication/tlschannel.cpp
+            ../../src/rootca/rootca.S
```
Comment (al1img) [inline tests/communication/CMakeLists.txt:96]: sort by name

File: tests/communication/CMakeLists.txt [hunk 20]
```diff
@@ -51,13 +83,35 @@ target_include_directories(app PRIVATE ${CMAKE_CURRENT_BINARY_DIR})
...
+    ${ZEPHYR_MBEDTLS_MODULE_DIR}/library/psa_crypto_driver_wrappers_no_static.c
+    ${ZEPHYR_MBEDTLS_MODULE_DIR}/library/psa_crypto.c TARGET_DIRECTORY mbedTLS PROPERTIES HEADER_FILE_ONLY ON
 )
+
+target_sources(
+    app PRIVATE ${aoscore_source_dir}/src/common/crypto/mbedtls/drivers/psa_crypto_driver_wrappers_no_static.c
+                ${aoscore_source_dir}/src/common/crypto/mbedtls/drivers/psa_crypto.c
+)
+
+target_include_directories(mbedTLS INTERFACE ${aoscore_source_dir}/src/common)
```
Comment (al1img) [inline tests/communication/CMakeLists.txt:116]: Is these includes required?

File: tests/communication/src/mocks/tls/certhandlermock.hpp [hunk 21]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef CERTHANDLERMOCK_HPP_
+#define CERTHANDLERMOCK_HPP_
+
+#include <aos/iam/certhandler.hpp>
+
+class MockCertHandler : public aos::iam::certhandler::CertHandlerItf {
```
Comment (al1img) [inline tests/communication/src/mocks/tls/certhandlermock.hpp:13]: CertHandlerMock

The idea was to reuse mock for different tests. Why you can't use existing `CertHandlerMock`?

File: tests/communication/src/mocks/tls/certloadermock.hpp [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+#include <mbedtls/pk.h>
+#include <mbedtls/rsa.h>
+#include <psa/crypto.h>
+
+#include <aos/common/crypto.hpp>
+#include <aos/common/cryptoutils.hpp>
+
+#include "rsaprivatekey.hpp"
+
+class MockCertLoader : public aos::cryptoutils::CertLoaderItf {
```
Comment (al1img) [inline tests/communication/src/mocks/tls/certloadermock.hpp:24]: Move to mocks folder.

File: tests/communication/src/tlschannel.cpp (hunk 17 above)
Comment (al1img) [reply tests/communication/src/tlschannel.cpp:24]: Please add headers

File: src/communication/tlschannel.hpp (hunk 18 above)
Comment (MykolaSolyanko) [reply src/communication/tlschannel.hpp:42]: Done

File: tests/communication/CMakeLists.txt (hunk 20 above)
Comment (MykolaSolyanko) [reply tests/communication/CMakeLists.txt:116]: no, removed

File: tests/communication/CMakeLists.txt (hunk 19 above)
Comment (MykolaSolyanko) [reply tests/communication/CMakeLists.txt:96]: Done

File: tests/communication/src/mocks/tls/certhandlermock.hpp (hunk 21 above)
Comment (MykolaSolyanko) [reply tests/communication/src/mocks/tls/certhandlermock.hpp:13]: i did not see that this mock already exist

File: tests/communication/src/mocks/tls/certloadermock.hpp (hunk 22 above)
Comment (MykolaSolyanko) [reply tests/communication/src/mocks/tls/certloadermock.hpp:24]: Done

File: tests/communication/src/mocks/certloadermock.hpp [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+            = aos::Array<uint8_t>(reinterpret_cast<const uint8_t*>(mCertChainPem.Get()), mCertChainPem.Size() + 1);
+
+        certChain->PushBack(cert);
+
+        return {certChain, aos::ErrorEnum::eNone};
+    }
+
+    aos::RetWithError<aos::SharedPtr<aos::crypto::PrivateKeyItf>> LoadPrivKeyByURL(const aos::String& url) override
+    {
+        aos::StaticString<4096>  mPKPem;
```
Comment (mykola-kobets-epam) [inline tests/communication/src/mocks/certloadermock.hpp:55]: aos::crypto::cPrivKeyPEMLen

**Review by mykola-kobets-epam — APPROVED** [review]

**Review by al1img — APPROVED** [review]

**Pull request merged**